src/pcp2zabbix/src/zbxpcp.3
'\"! tbl | mmdoc
'\"macro stdmacro
.\"
.\" Copyright (C) 2015 Marko Myllynen <myllynen@redhat.com>
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\"
.TH ZBXPCP 3 "PCP" "Performance Co-Pilot"
.SH NAME
\f3zbxpcp\f1 \- Zabbix Agent Loadable PCP Module
.SH DESCRIPTION
The
.B zbxpcp
loadable module extends Zabbix agent by making Performance Co-Pilot
(PCP) metrics available to it.
With the
.B zbxpcp
module configured in
.I zabbix_agentd.conf
all the PCP metrics are available from the Zabbix agent like any other
agent items. As a loadable module (DSO)
.B zbxpcp
does not rely on any external programs but directly uses the PCP APIs to
fetch PCP metrics when requested.
.PP
A typical PCP installation on Linux offers over 1000 metrics by default
and is in turn extensible with its own plugins, or PMDAs (``Performance
Metrics Domain Agents''). In addition to very complete
.I proc
based statistics, readily available PCP PMDAs provide support for such
system and application level components as Apache, CIFS, 389 Directory
Server, GFS2, Gluster, InfiniBand, KVM, MySQL, NFS, Postfix, PostgreSQL,
Samba, and Sendmail, among others. In addition to Linux, PCP also runs
on Mac OS X, FreeBSD, NetBSD, Solaris, and Windows.
.PP
For PCP introduction, see
.BR PCPIntro (5).
To investigate the available PCP metrics, use for example
.BR pminfo (1).
The
.I PCP Quick Reference Guide
is available at its home page http://pcp.io\:/.
.PP
For general information about Zabbix data collection and loadable
modules, see
https://www.zabbix.com/documentation/3.0/manual/config/items. For Zabbix
introduction and downloads, see http://www.zabbix.com/.
.PP
.B zbxpcp
is compatible with the Zabbix module API version
.IR 1 .
.SH CONFIGURATION
First make sure PCP is installed and configured properly, see the above
references for instructions and use for example
.BR pminfo (1)
to make sure the PCP metrics can be fetched. To enable the
.B zbxpcp
loadable module in a Zabbix agent, the following lines must be added to
the Zabbix agent configuration file
.I zabbix_agentd.conf
(make sure to use the correct
.IR LoadModulePath ):

.RS +4
.ft CW
.nf
LoadModulePath=/usr/lib64/zabbix/agent/
LoadModule=zbxpcp.so
.fi
.ft P
.RE

After restarting the Zabbix agent all the PCP metrics will be available
with the ``\c
.BR pcp. ''
prefix like all the other agent items. This can be verified with the
commands:

.RS +4
.ft CW
.nf
$ pminfo -t
$ zabbix_agentd -p
$ pminfo -f kernel.all.sysfork
$ zabbix_get -s 127.0.0.1 -p 10050 -k pcp.kernel.all.sysfork
.fi
.ft P
.RE
Then to collect PCP metrics from Zabbix agents, the Zabbix server side
configuration is as usual (see Zabbix documentation if needed).
.PP
The PCP
.B pmcd
service must always be running when starting up a
.B zbxpcp
enabled Zabbix agent, otherwise the module will fail to load and the PCP
metrics will not become available. Special care must be taken to make
sure this happens also when rebooting the system. A
.B pmcd
restart needs to be followed by a Zabbix agent restart.
.SH FILES
.PD 0
.TP 10
.BI /etc/zabbix/zabbix_agentd.conf
Typical Zabbix agent configuration file location.
.TP 10
.BI $(PCP_LIB_DIR)/zabbix/agent/zbxpcp.so
Zabbix agent loadable PCP module.
.PD
.SH SEE ALSO
.BR PCPIntro (1),
.BR pminfo (1),
.BR pmrep (1),
.BR zabbix_get (1)
and
.BR zabbix_agentd (8).


CHANGELOG
pcp-3.10.9 (16 December 2015)
    - Work in progress [see http://pcp.io/roadmap]

pcp-3.10.8 (30 October 2015)
    - pmdaslurm: new PMDA exporting HPC scheduler metrics
    - pmdalinux: correctly report high speed network link speeds
    - pmdalinux: support for wireless network interfaces
    - pmdalinux: add support for NVME devices
    - pmie: fqdn functionality, added %c for action strings
    - pmlogextract: runtime reducing instance optimizations
    - pmlogrewrite: metric name lex pattern change
    - pmlogger: change semantics for first logging operation
    - rc_pmlogger: shell escape for control file "directory" field
    - pcp-atop: update with latest atop features (esp. NFS)
    - pcp-atop: fix initial fetch time offset for the globals
    - pcp-atop: fix -b/-e options to match man page description
    - docs: update Quick Ref Guide with pcp-atop/pcp-atopsar
    - docs: remove outdated pmatop man page, see pcp-atop(1)
    - python api: fix local-context mode type error
    - python api: improve handling of non-ascii instance names
    - python api: simple debugging interface to access pmDebug
    - python api: ensure an interruptible sleep is used
    - python api: add single host/archive pmOption setter methods
    - libpcp: add $PCP_DEBUG to initialize pmDebug
    - libpcp: extend __pmAF* family with __pmAFsetup
    - libpcp: resolve false-context-sharing corner cases
    - pcp2graphite: provide a local-context mode option
    - pmmgr: add the subtarget-containers option
    - pmmgr: add pmlogreduce support
    - rc_pmcd: be more careful with auto-install/-remove
    - pmcd: allow dynamic switching of monitored containers
    - pmdapmcd: add pmcd.client.container diagnostic metric
    - pmdaroot: new metric mapping containers to their cgroups
    - pmdaroot: add cgroup heuristics for non-systemd setups
    - pmdaroot: fix timeliness of creation of socket connection
    - pmdads389: add normalized dn cache metrics
    - pmdads389: instantaneous vs discrete metric corrections
    - pmdads389: send correct error codes when not connected
    - pmdamounts: correct a number of 32-bit unsafe calculations
    - pmdanfsclient: improve PMDA error handling
    - pmdaperfevent: fix invalid metric names
    - pmdaperfevent: add reference clock cycles for NHM and WSM
    - docs: added upgrade instructions to pmdaperfevent man page
    - containers: bindmount /dev/log for syslog messages
    - build: fix FreeBSD 10.2 with dtrace probes auto-enabled

pcp-3.10.7 (16 September 2015)
    - containers: added the pcp-pmie container
    - containers: remove /proc bind mount, conflicts with --pid=host
    - pcp-atopsar: new PMAPI sar command based on http://atoptool.nl
    - pcp-uptime: fix sigsegv from missing metrics on first fetch
    - pmchart: fix memory corruption when restoring Saved Hosts list
    - pmchart: fix sigsegv opening a view when no archive available
    - rc_scripts: pmie/pmlogger - logged, no longer send cron email
    - rc_scripts: pmie/pmlogger - support for control.d directories;
      backward-compatible /etc/pcp/{pmie,pmlogger}/control extension
    - rc_scripts: pmcd/pmlogger - provide a localhost-only setting
    - pmdamic: new PMDA for Intel MIC card metrics
    - pmdaproc: export CPU scheduler cgroup metrics from cpu.stat
    - pmdalinux: export IPv6 kernel metrics from /proc/net/snmp6
    - pmdalinux: fix variable types for several disk metrics
    - pmdalinux: printf int64_t and/or uint64_t support (for softnet)
    - pmdajson: updates to allow use with python 2.6 installations
    - pmcpp: support inline shell command execution
    - pmwebd: multiple metrics/instances per log graphite mode speedup
    - libpcp: use public domain getdate instead of gnulib version
    - libpcp: optimise archive decompression for common operations
    - libpcp: revert pdubuf tsearch-based optimisation, fails on OSX
    - pmieconf: add dmthin pool utilization monitoring scripts
    - python: more care taken in pmOptions destructor (could sigsegv)
    - windows port: 64-bit format specifiers throughout code base
    - windows port: fix a libpcp dlsym bug and IPC on Windows
    - windows port: fix a python crash when free'ing resources
    - pmdawindows: fix mem.cache.bytes metric kernel extraction
    - perl pmda module: fix double free error in local_sock() routine
      (observed to cause pmdamemcache crash during error path exit)
    - pmdanginx: correct errors returned for not-running nginx daemon
    - pmdamounts: add mount fs usage metrics via statvfs(3)
    - pmdapostfix: improve error handling for files vs directories
    - pmdanfsclient: fix typos in metric value parsing routines
    - pmlogconf: remove tool-specific intervals from tool templates
    - pmlogconf: add nginx, elasticsearch and memcache configs
    - docs: update quickref guide for easier command pasting
    - docs: update references to init scripts, pmlogger control files
    - docs: update pmdanginx(1) describing nginx configuration needs
    - build: configure.ac preserves optional gcc flags in the build
    - build: cater for builds using the latest Vector (now 1.0.2)
    - build: fix LGPL license annotations in headers+libraries
    - build: current FreeBSD toolchain updates

pcp-3.10.6 (4 August 2015)
    - rc_scripts: use syslog over email for error notification
    - libpcp: catch bad valfmt encoding in interp.c (SGI BZ 1100)
    - libpcp: $PATH-alike lists for derived metrics configurations
    - libpcp: add instant() method for derived metrics
    - libpcp: fix locking problem in pmGetConfig(3) routine
    - python: fix bug in pmsubsys causing timestamp to reset.
    - misc: several Coverity scan bug fixes, across the board
    - containers: add version and arch to container image filenames
    - containers: add crontab entries for containers
    - containers: don't fix pcp user since pcp-base already does
    - pmlogcheck: teach it about valfmt badness

src/pcp2zabbix/src/zbxpcp.c
/*
 * Zabbix loadable PCP module
 *
 * Copyright (C) 2015 Marko Myllynen <myllynen@redhat.com>
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation; either version 2.1 of the License, or
 * (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 * License for more details.
 *
 * For description on Zabbix loadable modules, see:
 * https://www.zabbix.com/documentation/3.0/manual/config/items/loadablemodules
 */

/*
 * TODO
 * - derived metrics support
 * - config file support
 *   - conn type
 *   - conn target
 *   - replace defines
 *   - list of supported metrics
 *   - scaling
 */

#ifndef ZBX_PCP_METRIC_BASE
#define ZBX_PCP_METRIC_BASE ""
#endif

#ifndef ZBX_PCP_METRIC_PREFIX
#define ZBX_PCP_METRIC_PREFIX "pcp."
#endif

/* PCP includes.  */
#include "pmapi.h"
#include "impl.h"

/* Zabbix includes.  */
#include "module.h"

/*
 * PCP connection
 */
static int ctx = -1;

int zbx_module_pcp_init()
{
    ctx = pmNewContext(PM_CONTEXT_HOST, "localhost");
    return ctx;
}

int zbx_module_pcp_uninit()
{
    return pmDestroyContext(ctx);
}

/*
 * Zabbix connection
 */
int zbx_module_init()
{
    if (zbx_module_pcp_init() < 0)
        return ZBX_MODULE_FAIL;
    return ZBX_MODULE_OK;
}

int zbx_module_api_version()
{
    return ZBX_MODULE_API_VERSION_ONE;
}

static int metric_count = 0;
static ZBX_METRIC *metrics = NULL;
void zbx_module_pcp_add_metric(const char *name);

ZBX_METRIC *zbx_module_item_list()
{
    int sts;
    ZBX_METRIC *mptr;
    static ZBX_METRIC empty[] = { {NULL} };

    /* Add PCP metrics to the Zabbix metric set.  */
    sts = pmTraversePMNS(ZBX_PCP_METRIC_BASE, zbx_module_pcp_add_metric);
    if (sts < 0 || !metric_count) { free(metrics); return empty; }

    /* Finalize the Zabbix set.  */
    mptr = metrics;
    metrics = (ZBX_METRIC *)realloc(mptr, (metric_count + 1) * sizeof(ZBX_METRIC));
    if (metrics == NULL) { free(mptr); return empty; }
    metrics[metric_count].key = NULL;

    return metrics;
}

void zbx_module_item_timeout(int timeout)
{
    __pmSetRequestTimeout((double)timeout);
}

int zbx_module_uninit()
{
    if (zbx_module_pcp_uninit() != 0)
        return ZBX_MODULE_FAIL;
    return ZBX_MODULE_OK;
}

/*
 * Zabbix/PCP connection
 */
int zbx_module_pcp_fetch_metric(AGENT_REQUEST *request, AGENT_RESULT *result);

void zbx_module_pcp_add_metric(const char *name)
{
    int sts;
    pmID pmid[1];
    pmDesc desc[1];
    char *metric;
    unsigned flags = 0;
    char *param = NULL;
    int *instlist;
    char **namelist;
    ZBX_METRIC *mptr = metrics;

    /* PCP preparations.  */
    sts = pmLookupName(1, (char **)&name, pmid);
    if (sts < 0) return;
    sts = pmLookupDesc(pmid[0], desc);
    if (sts < 0) return;

    /* Construct the Zabbix metric name.  */
    metric = (char *)malloc(strlen(ZBX_PCP_METRIC_PREFIX) + strlen(name) + 1);
    if (metric == NULL) return;
    strcpy(metric, ZBX_PCP_METRIC_PREFIX);
    strcat(metric, name);

    /* Pick a PCP metric instance for use with zabbix_agentd -p.  */
    if (desc[0].indom != PM_INDOM_NULL) {
        sts = pmGetInDom(desc[0].indom, &instlist, &namelist);
        if (sts < 0) { free(metric); return; }
        if (sts) {
            flags = CF_HAVEPARAMS;
            param = strdup(namelist[0]);
            free(instlist);
            free(namelist);
        }
    }

    /* Ready for Zabbix.  */
    metrics = (ZBX_METRIC *)realloc(mptr, (metric_count + 1) * sizeof(ZBX_METRIC));
    if (metrics == NULL) { metrics = mptr; free(metric); free(param); return; }
    metrics[metric_count].key = metric;
    metrics[metric_count].flags = flags;
    metrics[metric_count].function = zbx_module_pcp_fetch_metric;
    metrics[metric_count].test_param = param;
    metric_count++;
}

int zbx_module_pcp_fetch_metric(AGENT_REQUEST *request, AGENT_RESULT *result)
{
    int sts;
    char *metric[] = { request->key + strlen(ZBX_PCP_METRIC_PREFIX) };
    char *inst;
    pmID pmid[1];
    pmDesc desc[1];
    pmResult *rp;
    pmAtomValue atom;
    int iid = 0;
    int i;

    /* Parameter is the instance.  */
    switch(request->nparam) {
        case 0:
            inst = NULL;
            break;
        case 1:
            inst = get_rparam(request, 0);
            break;
        default:
            SET_MSG_RESULT(result, strdup("Extraneous instance specification."));
            return SYSINFO_RET_FAIL;
            break;
    }

    /* Preparations and sanity checks.  */
    sts = pmLookupName(1, metric, pmid);
    if (sts < 0) return SYSINFO_RET_FAIL;
    sts = pmLookupDesc(pmid[0], desc);
    if (sts < 0) return SYSINFO_RET_FAIL;
    if (inst != NULL && desc[0].indom == PM_INDOM_NULL) {
        SET_MSG_RESULT(result, strdup("Extraneous instance specification."));
        return SYSINFO_RET_FAIL;
    }
    if ((inst == NULL && desc[0].indom != PM_INDOM_NULL) ||
        (request->nparam == 1 && !strlen(inst))) {
        SET_MSG_RESULT(result, strdup("Missing instance specification."));
        return SYSINFO_RET_FAIL;
    }
    if (desc[0].indom != PM_INDOM_NULL) {
        iid = pmLookupInDom(desc[0].indom, inst);
        if (iid < 0) {
            SET_MSG_RESULT(result, strdup("Instance not available."));
            return SYSINFO_RET_FAIL;
        }
    }

    /* Fetch the metric values.  */
    sts = pmFetch(1, pmid, &rp);
    if (sts < 0) return SYSINFO_RET_FAIL;
    if (rp->vset[0]->numval < 1) {
        pmFreeResult(rp);
        SET_MSG_RESULT(result, strdup("No value available."));
        return SYSINFO_RET_FAIL;
    }

    /* Locate the correct instance.  */
    for (i = 0; desc[0].indom != PM_INDOM_NULL && i < rp->vset[0]->numval; i++) {
        if (rp->vset[0]->vlist[i].inst == iid) {
            break;
        }
    }
    if (i == rp->vset[0]->numval) {
        pmFreeResult(rp);
        return SYSINFO_RET_FAIL;
    }

    /* Extract the wanted value.  */
    sts = pmExtractValue(rp->vset[0]->valfmt, &rp->vset[0]->vlist[i],
                         desc[0].type, &atom, desc[0].type);
    pmFreeResult(rp);
    if (sts < 0) return SYSINFO_RET_FAIL;

    /* Hand it to the caller.  */
    switch(desc[0].type) {
        case PM_TYPE_32:
            SET_UI64_RESULT(result, atom.l);
            break;
        case PM_TYPE_U32:
            SET_UI64_RESULT(result, atom.ul);
            break;
        case PM_TYPE_64:
            SET_UI64_RESULT(result, atom.ll);
            break;
        case PM_TYPE_U64:
            SET_UI64_RESULT(result, atom.ull);
            break;
        case PM_TYPE_FLOAT:
            SET_DBL_RESULT(result, atom.f);
            break;
        case PM_TYPE_DOUBLE:
            SET_DBL_RESULT(result, atom.d);
            break;
        case PM_TYPE_STRING:
            SET_STR_RESULT(result, strdup(atom.cp));
            break;
        default:
            SET_MSG_RESULT(result, strdup("Unsupported metric value type."));
            return SYSINFO_RET_FAIL;
            break;
    }

    /* Success.  */
    return SYSINFO_RET_OK;
}
